src/tsdb.c

```diff
@@ -225,86 +226,100 @@ int SeriesAddSample(Series *series, api_timestamp_t timestamp, double value) {
     }
 
     // No sample within range
-    if (res != CR_OK || sample->timestamp > iterator->maxTimestamp) {
-        return CR_END;   
+    if (res != CR_OK || sample->timestamp > iterator->maxTimestamp ||
+                        sample->timestamp < iterator->minTimestamp) {
+        return CR_END;
     } 
     return CR_OK;
 }
 
 void SeriesIteratorClose(SeriesIterator *iterator) {
-    if (iterator->chunkIterator != NULL)
-        iterator->series->funcs->FreeChunkIterator(iterator->chunkIterator);
-    if (iterator->dictIter != NULL)
-        RedisModule_DictIteratorStop(iterator->dictIter);
-    *iterator = (SeriesIterator){ 0 };
+    iterator->series->funcs->FreeChunkIterator(iterator->chunkIterator, iterator->reverse);
+    RedisModule_DictIteratorStop(iterator->dictIter);
 }
 
+// Fills sample from chunk. If all samples were extracted from the chunk, we 
+// move to the next chunk.
 ChunkResult SeriesIteratorGetNext(SeriesIterator *iterator, Sample *currentSample) {
-    Chunk_t *currentChunk = iterator->currentChunk;
+    ChunkResult res;
     ChunkFuncs *funcs = iterator->series->funcs;
+    Chunk_t *currentChunk = iterator->currentChunk;
 
-    ChunkResult res = funcs->ChunkIteratorGetNext(iterator->chunkIterator, currentSample);
+    res = iterator->GetNext(iterator->chunkIterator, currentSample);
     if (res == CR_END) { // Reached the end of the chunk
-        if (!RedisModule_DictNextC(iterator->dictIter, NULL, (void*)&iterator->currentChunk) ||
-            funcs->GetFirstTimestamp(currentChunk) > iterator->maxTimestamp) {
+        if (!iterator->DictGetNext(iterator->dictIter, NULL, (void*)&currentChunk) ||
+            funcs->GetFirstTimestamp(currentChunk) > iterator->maxTimestamp ||
+            funcs->GetLastTimestamp (currentChunk) < iterator->minTimestamp) {
             return CR_END;       // No more chunks or they out of range
         }
-        funcs->FreeChunkIterator(iterator->chunkIterator);
-        iterator->chunkIterator = funcs->NewChunkIterator(iterator->currentChunk);
-        funcs->ChunkIteratorGetNext(iterator->chunkIterator, currentSample);
+        funcs->FreeChunkIterator(iterator->chunkIterator, false);
+        iterator->chunkIterator = funcs->NewChunkIterator(currentChunk, false);
+        if(iterator->GetNext(iterator->chunkIterator, currentSample) != CR_OK) {
+            return CR_END;
+        }
     }
 
-    if (currentSample->timestamp > iterator->maxTimestamp) {
+    // check timestamp is within range
+    if (currentSample->timestamp < iterator->minTimestamp ||
+        currentSample->timestamp > iterator->maxTimestamp) {
         return CR_END;          // Reach end of range requested
     } 
     return CR_OK;
 }
 
 CompactionRule *SeriesAddRule(Series *series, RedisModuleString *destKeyStr, int aggType, uint64_t timeBucket) {
     CompactionRule *rule = NewRule(destKeyStr, aggType, timeBucket);
-    if (rule == NULL ) {
-        return NULL;
-    }
+    if (rule == NULL ) return NULL;
```

Comment:
@ashtul please revert this, I don't want the the ifs to become one-liners, its less readable when you have a big chunk of complex code that you need to understand.